android/app/src/main/java/com/zulipmobile/sharing/SharingModule.kt

```diff
@@ -12,6 +12,7 @@ internal class SharingModule(reactContext: ReactApplicationContext)
     @ReactMethod
     fun getInitialSharedContent(promise: Promise) {
         promise.resolve(initialSharedData)
+        initialSharedData = null
```

Comment:
Added a much more verbose commit message regarding this, I am also a bit skeptical about the placement of this code and have written about an alternative in the commit message itself, please share your views on it.

Reply:
commit ded9b49277d246ea9669465ad575edeccc37f59d

    sharing: Reset initialSharedData after sending the map to JS.
    
    This resolves the 'Map already consumed' bug.

(and much more detail, which is excellent, thank you!)

So effectively this solution makes `getInitialSharedContent` no longer a getter, but instead something that consumes its input. If we call it twice, the second time will get nothing, because the first call already consumed it.

I think that's *probably* the behavior we actually want in any case, even if there weren't a crash involved. When you take an action to share something, that's a single request to do so -- we don't want to be later picking it up from the middle and trying to act on it again.

Framed that way, I think this is the perfect place to implement that behavior -- right where the data is being handed off to the JS that's asked for it, we cross it off as having been consumed.

I have two remarks from there:
* Because this method now has an important side effect, it should have a name that sounds less like a plain old getter. Just replacing "get" with "read" might be good -- that makes the name similar to reading data from an input stream, which has exactly the same kind of side effect.
* There's another place in the existing code where we handle some data that has a lot in common with this structurally: when you start the app by opening a notification. That code currently has a different behavior:
  ```java
    @ReactMethod
    public void getInitialNotification(Promise promise) {
        if (null == initialNotification) {
            promise.resolve(null);
        } else {
            promise.resolve(Arguments.fromBundle(initialNotification));
        }
    }
  ```
  So if we call `getInitialNotification` several times, we get the same data each time. I think the `Arguments.fromBundle` there basically avoids this "Map already consumed" error by creating a new map on each call, as a copy of `initialNotification`.

  Could you try applying the same repro recipe with opening a notification, instead of sharing text? I am curious what the behavior is. From what you've discovered, I'm guessing that on the last step when you try starting the app from recent apps, we'll end up re-navigating to the conversation where the notification was -- which is a bug.

  If we do have that bug in notifications, then we should fix it by probably doing the same thing as this commit does, and that will also be helpful confirmation that this commit is the right direction for fixing the sharing issue. (Or if we turn out not to, then that will be interesting and I'll be curious to understand better what's different between the two cases.)

Oh and I guess also a small comment on the commit message:

    But it seems that react native discovers that the data in
    initialSharedIntent is outdated and hence is unsafe to use thereby
    giving us the exception and the bug.

I think it's not exactly that it's "outdated", so much as it is that the memory underlying that `WritableMap` has been re-used for handing that data over to the JS code.

Reply:
> Could you try applying the same repro recipe with opening a notification, instead of sharing text?

I did it and we end up re navigating to the conversation where the notification pointed.
Setting up `initialNotification` to null fixes it.

I should probably make a different PR to fix it.

Reply:
Sounds good -- please do! You might first file a quick issue for it.

Reply:
https://github.com/zulip/zulip-mobile/issues/4758